lib/lightning_web/live/credential_live/oauth_component.ex

```diff
@@ -0,0 +1,689 @@
+  end
+
+  def update(%{userinfo: userinfo}, socket) do
+    {:ok, socket |> assign(userinfo: userinfo, authorizing: false, error: nil)}
+  end
+
+  def update(%{scopes: scopes}, socket) do
+    handle_scopes_update(scopes, socket)
+  end
+
+  defp reset_assigns(socket) do
+    socket
+    |> assign_new(:userinfo, fn -> nil end)
+    |> assign_new(:authorizing, fn -> false end)
+  end
+
+  defp update_assigns(
+         socket,
+         %{
+           form: form,
+           id: id,
+           token_body_changeset: token_body_changeset,
+           update_body: update_body,
+           provider: provider
+         },
+         token
+       ) do
+    socket
+    |> assign(
+      form: form,
+      id: id,
+      token_body_changeset: token_body_changeset,
+      token: token,
+      error: token_error(token),
+      update_body: update_body,
+      provider: provider
+    )
+  end
+
+  defp update_client(socket) do
+    socket
+    |> assign_new(:client, fn %{token: token} ->
+      socket
+      |> build_client()
+      |> case do
+        {:ok, client} -> client |> Map.put(:token, token)
+        {:error, _} -> nil
+      end
+    end)
+    |> assign_new(:authorize_url, fn %{client: client} ->
+      if client do
+        socket.assigns.provider.authorize_url(
+          client,
+          build_state(socket.id, __MODULE__, socket.assigns.id),
+          []
+        )
+      end
+    end)
+  end
+
+  defp handle_code_update(code, socket) do
+    client = socket.assigns.client
+
+    # NOTE: there can be _no_ refresh token if something went wrong like if the
+    # previous auth didn't receive a refresh_token
+
+    case socket.assigns.provider.get_token(client, code: code) do
+      {:ok, client} ->
+        client.token
+        |> token_to_params()
+        |> socket.assigns.update_body.()
+
+        {:ok, socket |> assign(authorizing: false, client: client)}
+
+      {:error, %OAuth2.Response{status_code: 400, body: body}} ->
+        Logger.error("Failed retrieving token from provider:\n#{inspect(body)}")
+
+        send_update(self(), __MODULE__,
+          id: socket.assigns.id,
+          error: :token_failed
+        )
+
+        {:ok, socket}
+    end
+  end
+
+  defp handle_scopes_update(scopes, socket) do
+    authorize_url =
+      socket.assigns.provider.authorize_url(
+        socket.assigns.client,
+        build_state(socket.id, __MODULE__, socket.assigns.id),
+        scopes
+      )
+
+    {:ok, socket |> assign(authorize_url: authorize_url)}
+  end
+
+  @impl true
+  def handle_event("authorize_click", _, socket) do
+    {:noreply, socket |> assign(authorizing: true, userinfo: nil, error: nil)}
+  end
+
+  def handle_event("try_userinfo_again", _, socket) do
+    Logger.debug("Attempting to retrieve userinfo again...")
+
+    pid = self()
+    Task.start(fn -> get_userinfo(pid, socket) end)
+    {:noreply, socket |> assign(authorizing: true, error: nil, userinfo: nil)}
```

Comment:
Can you add a TODO here to change this over to use async_assign instead; this originally was written before that.

Reply:
Done